src/instrumenter/annotations.ts

```diff
@@ -546,6 +608,163 @@ export function gatherContractAnnotations(
+                const target = [...targets][0];
+
+                if (target instanceof FunctionDefinition && args.length > 0) {
+                    const params = target.vParameters.vParameters;
+
+                    mapArgsToParams(signature, args, params, localAliases);
+                }
+
+                for (const { expression, message } of properties) {
+                    const annotation = parseAnnotation(expression, target, compilerVersion);
+
+                    assert(
+                        annotation instanceof SProperty,
+                        `Only properties are allowed to be defined in macros. Unsupported annotation: ${expression}`
+                    );
+
+                    /**
+                     * Update variable names with supplied values from macro instruction.
+                     */
+                    annotation.expression.walk(walker);
+                    annotation.label = message;
+                    annotation.prefix = "#";
+
+                    /**
+                     * @todo This is temporary hack. Need to find a better way to do this.
+                     */
+                    const dummyDoc = new StructuredDocumentation(
+                        0,
+                        "0:0:0",
+                        "StructuredDocumentation",
+                        annotation.prefix + annotation.pp()
+                    );
+
+                    const metaToInject = new PropertyMetaData(
+                        dummyDoc,
+                        target,
+                        dummyDoc.text,
+                        annotation,
+                        0,
+                        dummyDoc.text
+                    );
+
+                    validateAnnotation(target, metaToInject);
+
+                    const metasToInject = injections.get(target);
+
+                    if (metasToInject === undefined) {
+                        injections.set(target, [metaToInject]);
+                    } else {
+                        metasToInject.push(metaToInject);
+                    }
+                }
+            }
+
+            /**
+             * Remove macro annotation as it will not pass type-checking.
+             * It is now represented by other annotations in `injections`,
+             * so there is no further use for macro annotation itself.
+             */
+            metas.splice(m, 1);
+        }
+    }
+
+    /**
+     * Merge injected annotations with original annotations.
+     */
+    for (const [target, metasToInject] of injections) {
+        const metas = annotationMap.get(target);
+
+        if (metas === undefined) {
+            annotationMap.set(target, metasToInject);
+        } else {
+            metas.push(...metasToInject);
+        }
+    }
+
+    /**
+     * @todo list:
+     * [TODO]   1. Source locations are needed for property maps.
+     * [TODO]   2. Figure out how to use macro variable types.
+     * [TODO]   3. Figure out if macros can use other macros.
```

Comment:
As a first iteration lets not allow this. I don't think its straight forward what that would look like. One way to potentially reason about this is to:

1. Limit macros to contract-level annotations only
2. Add "inheritance" amongst macros. Then for example you could have a hierarchy of `ERC20` macros, with increasingly more properties.

One potential issue with this approach is that the children inherit all the properties of the parents. I suspect there may be cases when the "inheriting" macro needs to invalidate properties of the parent. This needs more discussion, so for now lets leave it off the table.

Reply:
Deleted point 3 in https://github.com/ConsenSys/scribble/pull/96/commits/e57764f3acdbedbb98e4e3d041446f2210f42531. Currently only `SProperty` descendants are allowed (and `SMacro` is not one of them).